### PR TITLE
feat: add frontier evm

### PR DESCRIPTION
**Pull request by gregdhill**
Closes https://github.com/interlay/interbtc/issues/1098

## TODO

- [x] Verify all parameters make sense (most are taken from Centrifuge or Moonbeam)
- [x] Review signing logic (for wallet compatibility) and account conversion
- [x] Confirm required precompiles (included standard Ethereum ones)
- [x] Cleanup service (de-duplicate code, add SQL DB support, resolve `TransactionConverter`)
- [x] Configurable contract creation
- [x] Add EVM contract fixtures

Adds the following pallets to the `interlay` and `kintsugi` runtimes in addition to those required for RPC compatibility.

### `BaseFee` (`pallet_base_fee`)

Controls the price per gas unit, updates based on the weight of the previous block - decreases if below or increases if above the target.

### `Ethereum` (`pallet_ethereum`)

Emulation for block processing, validates and stores (EVM) transactions then constructs the Ethereum compatible block `on_finalize`.

### `EVM` (`pallet_evm`)

Contains the `SputnikVM` to execute EVM bytecode compatible with the Ethereum mainnet.

#### Configuration

- Account version uses [`HashedAddressMapping`](https://github.com/paritytech/frontier/blob/1a718546085be20ce381b70e1f9e4c8b4d4b1f03/frame/evm/src/lib.rs#L715)

> All non-root calls are disabled.

### `EVMChainId` (`pallet_evm_chain_id`)

Stores the chain ID which is used by the EVM runner.

### CODE REVIEW

**Comment by gregdhill** [comment]
How the block / tx processing logic works:

1. User submits ethereum transaction over RPC https://github.com/paritytech/frontier/blob/bd5bd374fd50eb30eb381148243c46d882e057d5/client/rpc-core/src/eth.rs#L240-L247
2. Runtime overrides extrinsic types https://github.com/paritytech/frontier/blob/bd5bd374fd50eb30eb381148243c46d882e057d5/template/runtime/src/lib.rs#L468-L473 (adds additional signature and validity checks https://github.com/paritytech/frontier/blob/bd5bd374fd50eb30eb381148243c46d882e057d5/primitives/self-contained/src/checked_extrinsic.rs#L36)
3. On execution it will append that transaction, status and execution receipt to `Pending` https://github.com/paritytech/frontier/blob/bd5bd374fd50eb30eb381148243c46d882e057d5/frame/ethereum/src/lib.rs#L673
4. When the current Substrate block is built `on_finalize` it will include those transactions in an ethereum block https://github.com/paritytech/frontier/blob/bd5bd374fd50eb30eb381148243c46d882e057d5/frame/ethereum/src/lib.rs#L442
5. That block can be fetched over RPC using the Runtime API https://github.com/paritytech/frontier/blob/bd5bd374fd50eb30eb381148243c46d882e057d5/template/runtime/src/lib.rs#L828C21-L828C33

**Review by sander2 — COMMENTED** [review]
Looks ok to me, but honestly it's very difficult to review, because there's so much code involved.. 

One question I do have is, how can the evm interact with our runtime? E.g. if it wants to check a bitcoin inclusion?

File: parachain/runtime/interlay/src/evm.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+pub use runtime_common::evm::{
+    precompiles::InterBtcPrecompiles, BaseFeeThreshold, GetTransactionAction, SetEvmChainId, WEIGHT_PER_GAS,
+};
+
+pub type Precompiles = InterBtcPrecompiles<Runtime>;
+pub type AccountConverter = HashedAddressMapping<BlakeTwo256>;
+
+parameter_types! {
+    pub DefaultBaseFeePerGas: U256 = U256::from(CENTS);
```
Comment (gregdhill) [inline parachain/runtime/interlay/src/evm.rs:25]: The "flipper" contract fixture costs 133751 gas to create which would total `(133751 * (10**10 / 100)) / 10**10 = 1337.51 INTR ($19.7)`.

File: parachain/runtime/kintsugi/src/evm.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+pub use runtime_common::evm::{
+    precompiles::InterBtcPrecompiles, BaseFeeThreshold, GetTransactionAction, SetEvmChainId, WEIGHT_PER_GAS,
+};
+
+pub type Precompiles = InterBtcPrecompiles<Runtime>;
+pub type AccountConverter = HashedAddressMapping<BlakeTwo256>;
+
+parameter_types! {
+    pub DefaultBaseFeePerGas: U256 = U256::from(CENTS * 10);
```
Comment (gregdhill) [inline parachain/runtime/kintsugi/src/evm.rs:25]: The "flipper" contract fixture costs 133751 gas to create which would total `(133751 * (10**12 / 30_000 * 10)) / 10**12 = 44.58 KINT ($16.94)`.

File: parachain/runtime/common/src/evm/mod.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+};
+use pallet_ethereum::{Transaction, TransactionAction};
+use sp_core::Get;
+use sp_runtime::Permill;
+use sp_std::marker::PhantomData;
+
+pub mod precompiles;
+
+/// Current approximation of the gas/s consumption (Moonbeam)
+pub const GAS_PER_SECOND: u64 = 40_000_000;
```
[the remark this answers is not in the document]
Comment (gregdhill) [reply parachain/runtime/common/src/evm/mod.rs:13]: I'm actually not sure, but this is where Moonbeam defines it: https://github.com/moonbeam-foundation/moonbeam/blob/c55870bdda5bd40e97daffc27a5e3e1c037c5da9/runtime/moonbeam/src/lib.rs#L365-L369

[the remark this answers is not in the document]
Comment (gregdhill) [reply parachain/runtime/common/src/evm/mod.rs:13]: AFAIK that is used in production and Centrifuge also copied that for their runtime.

**Comment by gregdhill** [comment]
> how can the evm interact with our runtime?

We need to add precompiles, see for example Acala's [`MultiCurrencyPrecompile`](https://github.com/AcalaNetwork/Acala/blob/27a6d5ac1f7481bb4e16b1ea697db7b8e2210c8c/runtime/common/src/precompile/multicurrency.rs#L50).

**Review by sander2 — APPROVED** [review]

**Pull request merged**